src/python/zquantum/qaoa/recursive_qaoa.py

```diff
@@ -0,0 +1,247 @@
+
+            estimation_preprocessors = [partial(allocate_shots_uniformly, number_of_shots=1000)]
+            estimation_tasks_factory_generator = partial(
+                substitution_based_estimation_tasks_factory,
+                estimation_preprocessors=estimation_preprocessors
+            )
+            cost_function_factory = partial(
+                create_cost_function,
+                backend=QuantumBackend,
+                estimation_method=estimate_expectation_values_by_averaging,
+                parameter_preprocessors=None,
+            )
+
+            initial_params = np.array([0.42, 4.2])
+            optimizer = ...
+
+            recursive_qaoa = RecursiveQAOA(
+                3,
+                ansatz,
+                initial_params,
+                optimizer,
+                estimation_tasks_factory,
+                cost_function_factory,
+            )
+
+            solutions = RecursiveQAOA(cost_hamiltonian)
+        """
+
+        self._n_c = n_c
+        self._ansatz = ansatz
+        self._initial_params = initial_params
+        self._optimizer = optimizer
+        self._estimation_tasks_factory = estimation_tasks_factory
+        self._cost_function_factory = cost_function_factory
+
+    def __call__(
+        self,
+        cost_hamiltonian: IsingOperator,
+        qubit_map: Optional[List[List[int]]] = None,
+    ) -> List[Tuple]:
+        """Args:
+            cost_hamiltonian: Hamiltonian representing the cost function.
+            qubit_map: A list that maps qubits in reduced Hamiltonian back to original qubits, used for
+                subsequent recursions. (Not for the first recursion.)
+                Example:
+                    `qubit_map = [(2, -1), (3, 1)]`
+                        Indice of each tuple is the original qubit indice.
+                        1st term of tuple is qubit the index of tuple to be mapped onto,
+                        2nd term is if it will be mapped onto the same value or opposite of the qubit it
+                            is being mapped onto.
+                        In the above qubit_map, the original qubit 0 is now represented by the opposite
+                            value of qubit 2, and the original qubit 1 is now represented by the value of
+                            qubit 3.
+
+        Returns:
+            List[Tuple[int]] The solution(s) to recursive QAOA as a list of tuples; each tuple is a tuple
+                of integer bits.
+        """
+
+        n_qubits = count_qubits(change_operator_type(cost_hamiltonian, QubitOperator))
+
+        assert self._n_c < n_qubits
+
+        if not qubit_map:
+            qubit_map = []
+            for i in range(n_qubits):
+                qubit_map.append([i, 1])
+
+        # TODO: is there a better way to allow ansatzes to be modular? Not all
+        # ansatzes have a `cost_hamiltonian` attribute.
+        # You can't `partial` a constructor, right?
+        ansatz = copy(self._ansatz)
```

Comment:
You could pass ansatz factory instead of ansatz, but I guess that would be too meta.
I think the assumption that your QAOA ansatz will have `cost_function` field is a reasonable one.